examples/vision/super_resolution_sub_pixel.py

```diff
@@ -0,0 +1,513 @@
+outputs = tf.nn.depth_to_space(x, upscale_factor)
+
+model = tf.keras.Model(inputs, outputs)
+```
+
+```
+Model: "model"
+_________________________________________________________________
+Layer (type)                 Output Shape              Param #
+=================================================================
+input_1 (InputLayer)         [(None, 100, 100, 1)]     0
+_________________________________________________________________
+conv1 (Conv2D)               (None, 100, 100, 64)      1664
+_________________________________________________________________
+conv2 (Conv2D)               (None, 100, 100, 64)      36928
+_________________________________________________________________
+conv3 (Conv2D)               (None, 100, 100, 32)      18464
+_________________________________________________________________
+conv4 (Conv2D)               (None, 100, 100, 9)       2601
+_________________________________________________________________
+tf_op_layer_DepthToSpace (Te [(None, 300, 300, 1)]     0
+=================================================================
+Total params: 59,657
+Trainable params: 59,657
+Non-trainable params: 0
+```
+
+"""
+
+"""
+### Prepare train and test data sequences.
+"""
+
+train_gen = BSDS500(
+    input_img_size=input_img_size,
+    target_img_size=target_img_size,
+    crop_size=crop_size,
+    input_img_paths=input_img_paths,
+    transform_input=transform_input,
+    transform_target=transform_target,
+    batch_size=batch_size,
+    upscale_factor=upscale_factor,
+    channels=channels,
+)
+
+valid_gen = BSDS500(
+    input_img_size=input_img_size,
+    target_img_size=target_img_size,
+    crop_size=crop_size,
+    input_img_paths=valid_img_paths,
+    transform_input=transform_input,
+    transform_target=transform_target,
+    batch_size=batch_size,
+    upscale_factor=upscale_factor,
+    channels=channels,
+)
+
+"""
+## Train the model with custom training loop.
+"""
+
+model = get_model(upscale_factor=upscale_factor, channels=channels)
+loss_fn = tf.keras.losses.MeanSquaredError()
+optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)
+
+import math
+import numpy as np
+
+epochs = 50
+
+best_psnr = 0
+for epoch in range(epochs):
+    print("\nStart of epoch %d" % (epoch + 1,))
+    total_psnr = 0
+    total_loss = 0
+    validation_loss = 0
+    # Iterate over the batches of the dataset.
+    for step, (x_batch_train, y_batch_train) in enumerate(train_gen):
+        with tf.GradientTape() as tape:
+            logits = model(x_batch_train, training=True)
+            loss_value = loss_fn(y_batch_train, logits)
+        grads = tape.gradient(loss_value, model.trainable_weights)
+        optimizer.apply_gradients(zip(grads, model.trainable_weights))
+        total_loss += loss_value
+        print(
+            "Training loss (for one batch) at step %d: %.4f" % (step, float(loss_value))
+        )
+    print(
+        "Avg. Loss for each epoch %d: %.4f" % (epoch + 1, total_loss / len(train_gen))
```

Comment:
You can replace all of this with `model.fit()`, much nicer (and feature-complete, since you can now train on multiple GPUs or TPUs with no code change)